### DISCUSSION TITLE
Syntax for type casting

### DISCUSSION THREAD

**Discussion by amshaikh** [dotnet/csharplang#3902] [General]
The current syntax for typecast requires unnecessary bracketing eg:

((Myclass)(x.obj1.obj2)).data

(int)(3.14+xyz)

if we leave this pathetic syntax of c/c++ typecasting and adopt function like syntax, the above can easily be written as:

Myclass(x.obj1.obj2).data

**Comment by theunrepentantgeek** [comment]
Three observations ...

... generally speaking, new language features need to offer more value than simply providing an alternative syntax for something that's already possible. 

... function style syntax was well known at the time C# was first designed - in fact, it was used in Delphi's Object Pascal language, designed by Anders Hejlsberg, the original language designer of C#. While I don't know what they were, I'm sure the designers had good reasons for sticking with the C style

... introducing this style of cast would be a breaking change - the LDM is generally very reluctant to take on breaking changes unless they provide a great deal of value.

**Comment by YairHalberstadt** [comment]
You can define a method like this to allow fluent casting:

```csharp
public static T As<T>(object o) where T : class => (T)o;
```
It doesn't work for structs, or cast operators.

**Comment by Flutterish** [comment]
I find `( x.obj1.obj2 as MyClass ).data` and `x.obj1.obj2.As<MyClass>().data` more readable than `MyClass( x.obj1.obj2 ).data` which almost looks like a `new` call, not to mention in most situations you will want to save `x.obj1.obj2 as MyClass` as a local variable.

**Reply by PathogenDavid** [reply]
The problem with using `as` when you really want a cast is rather than an `InvalidCastException` (which states the types involved) you get a `NullReferenceException` (which is vague and unuseful if you don't have a debugger attached.)

This also [doesn't play nice with nullable reference types](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAxAOwK4A2BEwBcABHDiWQLABQDAAgMzlMBM5AsgJ4DCxKFAYBvBuUns2TAIwAGcgBEIMCOQC85AEQBlAPZgKAE1URtAbglTrk1u1kA2B4oDicGCrUAKfcABWcADGMOT6AJS2UpIaAHzkvuTQPAJCUOEAdF4QFuQA9HnkAO4QCDgAljgA5uT8ugAcjvIcIMqIcABm7ThBFPodSeQADvrC5aS85PhE5Aid3b0ZDAC+DEA==).
